core/matrix/coo.cpp

```diff
@@ -77,37 +77,82 @@ GKO_REGISTER_OPERATION(outplace_absolute_array,
 template <typename ValueType, typename IndexType>
 void Coo<ValueType, IndexType>::apply_impl(const LinOp *b, LinOp *x) const
 {
-    using Dense = Dense<ValueType>;
-    this->get_executor()->run(coo::make_spmv(this, as<Dense>(b), as<Dense>(x)));
+    using ComplexDense = Dense<to_complex<ValueType>>;
+
+    if (dynamic_cast<const Dense<ValueType> *>(b)) {
+        this->get_executor()->run(coo::make_spmv(this, as<Dense<ValueType>>(b),
+                                                 as<Dense<ValueType>>(x)));
+    } else {
+        auto dense_b = as<ComplexDense>(b);
+        auto dense_x = as<ComplexDense>(x);
+        this->get_executor()->run(coo::make_spmv(
+            as<Coo<remove_complex<ValueType>, IndexType>>(this),
+            dense_b->view_as_real().get(), dense_x->view_as_real().get()));
```

Comment:
You could also write `this->apply(dense_b->view_as_real().get(), dense_x->view_as_real().get())`, not sure this is the cleanest solution, though. We should also keep in mind what the error case would be here, so what kind of error message does the user get when this fails due to the wrong type? Before, they were warned about ValueType, suddenly they are being warned about `to_complex<ValueType>`, might be slightly confusing.

Reply:
I like it!